### PR TITLE
feat(decorators): validate decorators against model

### LINKED ISSUE #391: Decorator Validation
# Feature Request 🛍️

The current decorator support in Concerto has some issues which were deferred initially due to time constraints.
1. Decorators are not modelled types, rather they have a generic structure
2. Decorators cannot be defined within a namespace and imported / shared 

## Use Case

Allow users to define sets of decorators and to import and share them. Ensure that decorator arguments are specified and valid.

## Possible Solution

Define a root decorator concept that decorator definitions must extend.

Abstract decorator concept is added to the root (system) namespace:

```
namespace concerto.decorator@1.0.0

abstract concept Decorator {}
```

Users may specialise the decorator in their own namespaces:

```
namespace org.acme.decorators@1.0.0

import concerto.decorator@1.0.0.{Decorator}

concept Hide extends Decorator {
   o Boolean visibility
   o String docs optional
}
```

Users may use the defined decorator in their own namespaces:

```
namespace org.acme.person@1.0.0

import org.acme.decorators@1.0.0.{Hide}

concept Person {
   @Hide
   o String ssn
}
```

This declares a decorator called `Hide` that takes a single Boolean argument (the name of the property is ignored).

Update the `Decorator.validate`method to optionally validate all the decorators on namespaces/declarations/properties.
- Decorator names and imports are validated
- Decorator arguments are validated

## Context

Have a need for a rich set of decorators that can be managed over time.

## Detailed Description

The proposed change would NOT impact the Concerto parser or metamodel or break code generation.

### PR DESCRIPTION
**Pull request by dselman**
# Closes #391 

Adds a new option to the model manager to validate the names and structure of decorators and their arguments:
- Decorator must have been defined
- Arguments must match expected types (String,Integer,Long,Double,Boolean)
- Optional arguments are supported
- Arguments that are a concept reference must be assignable

### Changes

Documentation updates: https://github.com/accordproject/concerto-docs/pull/61

Using this Model Manager configuration:

```
validatedModelManager = new ModelManager({
            strict: true, decoratorValidation: {
                missingDecorator: 'error',
                invalidDecorator: 'error'
            }
        });
```

This file will error, because it uses the undeclared decorator `@Hide`:

```
namespace test@1.0.0
concept Person {
   o String name
   @Hide
   o String ssn
   @Hide(false)
   o Integer age
}
```

While this file will not:

```
namespace test@1.0.0
import concerto.decorator@1.0.0.Decorator

concept Hide extends Decorator {
    o Boolean hidden optional
}

concept Person {
   o String name
   @Hide
   o String ssn
   @Hide(false)
   o Integer age
}
```

### Flags
- Default behaviour should be unchanged (no validation)
- DecoratorManager.decorateModels has a new option `defaultNamespace` to support decorating model managers with command sets when the model manager `decoratorValidation` option is ON. See below.
- A new "system" model file has been created `concerto.decorator@1.0.0` that defines the abstract `Decorator` type.

The Decorator Command Set model has been updated to support an optional `namespace` property for `DecoratorCommand.decorator` (in addition to the existing decorator name). This is required so that the DecoratorManager can create imports for the decorators that are being added to a model file. If the namespace property is missing on a decorator command then the `defaultNamespace` option passed to DecoratorManager is used. If both are null and a decorator is applied, and decoratorValidation is ON then the model file will be flagged by the model manager as invalid.

### Related Issues
- Issue #<NUMBER>
- Pull Request #<NUMBER>

### Author Checklist
- [ ] Ensure you provide a [DCO sign-off](https://github.com/probot/dco#how-it-works) for your commits using the `--signoff` option of git commit.
- [ ] Vital features and changes captured in unit and/or integration tests
- [ ] Commits messages follow [AP format](https://github.com/accordproject/techdocs/blob/master/DEVELOPERS.md#commit-message-format)
- [ ] Extend the documentation, if necessary
- [ ] Merging to `main` from `fork:branchname`

### CODE REVIEW

File: packages/concerto-core/lib/rootmodel.js [hunk 1]
```diff
@@ -32,7 +32,10 @@ function getRootModel(versioned) {
     abstract concept Participant identified {}
     abstract concept Transaction {}
     abstract concept Event {}
-    `;
+    abstract concept Decorator {}
+    concept DotNetNamespace extends Decorator {
+       o String namespace
+    }`;
```
Comment (mttrbrts) [inline packages/concerto-core/lib/rootmodel.js:38]: Adding declarations to the root model without bumping the version?

File: packages/concerto-core/lib/rootmodel.js [hunk 2]
```diff
@@ -32,7 +32,10 @@ function getRootModel(versioned) {
     abstract concept Participant identified {}
     abstract concept Transaction {}
     abstract concept Event {}
-    `;
+    abstract concept Decorator {}
+    concept DotNetNamespace extends Decorator {
```
Comment (mttrbrts) [inline packages/concerto-core/lib/rootmodel.js:36]: Does `DotNetNamespace` belong here? Perhaps in another namespace?

File: packages/concerto-core/test/data/decoratorcommands/test.cto [hunk 3]
```diff
@@ -1,7 +1,11 @@
 namespace test@1.0.0
+import concerto@1.0.0.Decorator
```
Comment (mttrbrts) [inline packages/concerto-core/test/data/decoratorcommands/test.cto:2]: Interesting, should this be implicitly in-scope? For comparison, what's the behaviour for `Concept`?
e.g. `concept X extends Concept`

File: packages/concerto-core/lib/rootmodel.js (hunk 1 above)
Comment (dselman) [reply packages/concerto-core/lib/rootmodel.js]: Moved this to a new ns.

File: packages/concerto-core/lib/rootmodel.js (hunk 2 above)
Comment (dselman) [reply packages/concerto-core/lib/rootmodel.js]: Moved to new ns.

File: packages/concerto-core/lib/basemodelmanager.js [hunk 4]
```diff
@@ -762,8 +781,10 @@ class BaseModelManager {
     fromAst(ast) {
         this.clearModelFiles();
         ast.models.forEach( model => {
-            const modelFile = new ModelFile( this, model );
-            this.addModelFile( modelFile, null, null, true );
+            if(model.namespace !== 'concerto@1.0.0') { // excludes the Concerto namespace, already added
```
[the remark this answers is not in the document]
Comment (dselman) [reply packages/concerto-core/lib/basemodelmanager.js]: Can't see that...

File: packages/concerto-core/lib/basemodelmanager.js [hunk 5]
```diff
@@ -83,6 +94,9 @@ class BaseModelManager {
      * @param {boolean} [options.addMetamodel] - When true, the Concerto metamodel is added to the model manager
      * @param {boolean} [options.enableMapType] - When true, the Concerto Map Type feature is enabled
      * @param {boolean} [options.importAliasing] - When true, the Concerto Aliasing feature is enabled
+     * @param {object} [options.decoratorValidation] - the decorator validation configuration
+     * @param {string} [options.decoratorValidation.defined] - the validation log level for defined decorators: off, warning, error
```
Comment (sanketshevkar) [inline packages/concerto-core/lib/basemodelmanager.js:98]:
```suggestion
     * @param {string} [options.decoratorValidation.missingDecorator] - the validation log level for defined decorators: off, warning, error
```

File: packages/concerto-core/lib/basemodelmanager.js [hunk 6]
```diff
@@ -83,6 +94,9 @@ class BaseModelManager {
...
+     * @param {string} [options.decoratorValidation.undefined] - the validation log level for undefined decorators: off, warning, error
```
Comment (sanketshevkar) [inline packages/concerto-core/lib/basemodelmanager.js:99]:
```suggestion
     * @param {string} [options.decoratorValidation.invalidDecorator] - the validation log level for invalidDecorator decorators: off, warning, error
```

**Pull request merged**